<a href="https://colab.research.google.com/github/rahuldey5156/Machine-Learning-using-Python/blob/main/Na%C3%AFve_Bayes_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import the necessary Python packages

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Let's download the iris dataset from its weblink as follows:

In [3]:
path = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

Assign the column names to the dataset as follows:

In [4]:
headernames = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'Class']

Next, we read the dataset using the Pandas data frame

In [5]:
data = pd.read_csv(path, names=headernames)
data.shape

(150, 5)

Observe the data

In [6]:
data.head()

sepal-length  sepal-width  petal-length  petal-width        Class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

Preprocess the data

In [7]:
X = data.iloc[:, :-1].values
y = data.iloc[:, 4].values

Divide the data into train and test splits, with 70% of the training data and 30% of the testing data

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

Scale the data

In [9]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

Train the model with the help of the GaussianNB class of sklearn as follows:

In [10]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train);

Now, we have to make the prediction, which can be done in the following way:

In [11]:
y_pred = classifier.predict(X_test)

Let's print the results as follows:

In [12]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
result = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(result)
result1 = classification_report(y_test, y_pred)
print("Classification Report:",)
print (result1)
result2 = accuracy_score(y_test,y_pred)
print("Accuracy:",result2)

Confusion Matrix:
[[20  0  0]
 [ 0  9  3]
 [ 0  0 13]]
Classification Report:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        20
Iris-versicolor       1.00      0.75      0.86        12
 Iris-virginica       0.81      1.00      0.90        13

       accuracy                           0.93        45
      macro avg       0.94      0.92      0.92        45
   weighted avg       0.95      0.93      0.93        45

Accuracy: 0.9333333333333333


Using the Naive Bayes classifier, we have achieved an accuracy of 93.33%